### DISCUSSION TITLE
Stremio returns 415 Media Unsupported code

### DISCUSSION THREAD

**Discussion by milouk** [n0bodysec/docker-images#3] [General]
Hello, 

Im using your image to host a personal stremio server over http. However im getting a 415 Media Unsupported error code when trying to stream anything. Do you have any idea what could be wrong?

Thanks

**Comment by snowyyd** [comment]
Hello,

are you using the latest commit ([#18b9295](https://github.com/n0bodysec/docker-images/commit/18b9295e43d0e08b32dbc2b0721abaeb539d26fc))?

What browser are you using?
Are you using a local IP/hostname to access the server/client?

---

**EDIT:** I was able to reproduce the issue. Apparently it is related to stremio-web itself (currently under development, not recommended for production builds). The same problem is present in the netlify deployment (stremio-development.netlify.app). The strange thing about this is that I can't reproduce the problem in all movies.

I can run the native Stremio app with the dockerized server without issues.

---

**EDIT:** Maybe it's related to the way both apps manage to get the m3u8 file.

Stremio Web: http://hostname.local:11470/hlsv2/806423f72f2e69b996ad9abed0ffccfe/master.m3u8?mediaURL=http%3A%2F%2Fhostname.local%3A11470%2Ff82f1ab37bb1c8c46a61ef1f6052e98284f2a51b%2F0

Stremio Core (app.strem.io): http://local.strem.io:11470/f82f1ab37bb1c8c46a61ef1f6052e98284f2a51b/0/hls.m3u8

"mediaURL" arg isn't even present in `server.js`.

**Reply by snowyyd** [reply]
Apparently this issue is related to the [stremio-video repository](https://github.com/Stremio/stremio-video).

**Reply by snowyyd** [reply]
Ok, actually I made a "fix" by editing only the `server.js` file (this file isn't the problem because stremio-core works well).

I managed to "bypass" all the `hlsv2` related queries to the "old queries".

Explanation:

Server file: https://dl.strem.io/four/master/server.js

Check this chunk of code (extracted from server.js):

```js
var setHLSFrom = function(req, res, next) {
    var first = req.params.first, second = req.params.second, from = req.params.from || req.query.from;
    if (from) req.params.from = decodeURIComponent(from);
    else if (first.length == 40) req.params.from = enginefs.baseUrlLocal+"/"+first+"/"+second;
    else if (first == "file" || first == "url") req.params.from = second;
    next();
};

enginefs.router.get("/:first/:second/hls.m3u8", setHLSFrom, HLS.masterPlaylistMiddleware);
enginefs.router.get("/:first/:second/master.m3u8", setHLSFrom, HLS.masterMultiPlaylistMiddleware);
```
\
I emulated the "old queries" with the following conditional in `setHLSFrom`

```js
if (req.params.first === 'hlsv2') {
	let tmp0 = req.query.mediaURL.split('/');
	req.params.first = tmp0[3];
	req.params.second = tmp0[4];
}
```
\
I have also removed the `masterMultiPlaylistMiddleware` function. This function controls the requests made to the `master.m3u8` file. Since the stremio-core make requests to `hls.m3u8` file, I replaced the executed callback for the `router.get` of the  `hls.m3u8` file.

```js
enginefs.router.get("/:first/:second/hls.m3u8", setHLSFrom, HLS.masterPlaylistMiddleware);
enginefs.router.get("/:first/:second/master.m3u8", setHLSFrom, HLS.masterPlaylistMiddleware); // original: HLS.masterMultiPlaylistMiddleware
```
\
At this point, the requests made to `master.m3u8` and `hls.m3u8` are identical.

Now I just need to perform all those fixes automatically

```Dockerfile
RUN sed -i -E "s|(var first = req\.params\.first)|if (req.params.first === 'hlsv2') { req.params.first = req.query.mediaURL.split('/')[3]; req.params.second = req.query.mediaURL.split('/')[4]; }\n    \1|" server/server.js
RUN sed -i -E 's/HLS.masterMultiPlaylistMiddleware\)/HLS.masterPlaylistMiddleware\)/' server/server.js
```
\
This method is very hacky in many ways. I don't know if this affect some movies/series in any way.
I also think that the problem is in the source code of `stremio-video`, so the fix should be done there.

**Reply by milouk** [reply]
wow, that was a quick fix. As per my investigation you are correct. I tested it and it works :) thanks again. Ill keep an eye on stremio-video for a proper fix. Cheers

**Reply by snowyyd** [reply]
You are welcome!

I still need to investigate further, I want to fix it in a proper way 😅
